File: php-advanced/src/Services/UserService.php
<?php

namespace App\Services;

use App\Models\User;
use App\Utils\Logger;
use App\Utils\Validator;
use Exception;

class UserService
{
    private array $users = [];
    private int $nextId = 1;
    private Logger $logger;
    private Validator $validator;

    public function __construct()
    {
        $this->logger = new Logger();
        $this->validator = new Validator();
    }

    public function createUser(string $name, string $email): User
    {
        if (!$this->validator->isValidName($name)) {
            $this->logger->error("Invalid name provided", ['name' => $name]);
            throw new Exception("Invalid name: must be between 2 and 100 characters");
        }

        if (!$this->validator->isValidEmail($email)) {
            $this->logger->error("Invalid email provided", ['email' => $email]);
            throw new Exception("Invalid email format");
        }

        $user = new User($this->nextId, $name, $email);
        $this->users[$this->nextId] = $user;
        $this->nextId++;

        $this->logger->info("User created successfully", ['user_id' => $user->getId()]);
        return $user;
    }

    public function getUser(int $id): ?User
    {
        return $this->users[$id] ?? null;
    }

    public function getAllUsers(): array
    {
        return array_values($this->users);
    }
}

File: java-advanced/src/main/java/functions/services/UserService.java
package functions.services;

import functions.models.User;
import functions.utils.Logger;
import functions.utils.Validator;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class UserService {
    private Map<Integer, User> users = new HashMap<>();
    private int nextId = 1;
    private Logger logger = new Logger();
    private Validator validator = new Validator();

    public User createUser(String name, String email) throws Exception {
        if (!validator.isValidName(name)) {
            logger.error("Invalid name provided", "{\"name\":\"" + name + "\"}");
            throw new Exception("Invalid name: must be between 2 and 100 characters");
        }

        if (!validator.isValidEmail(email)) {
            logger.error("Invalid email provided", "{\"email\":\"" + email + "\"}");
            throw new Exception("Invalid email format");
        }

        User user = new User(nextId, name, email);
        users.put(nextId, user);
        nextId++;

        logger.info("User created successfully", "{\"user_id\":" + user.getId() + "}");
        return user;
    }

    public User getUser(int id) {
        return users.get(id);
    }

    public List<User> getAllUsers() {
        return new ArrayList<>(users.values());
    }
}
